source/transactions-convenient-api/tests/commit-transienttransactionerror.yml

```diff
@@ -0,0 +1,202 @@
+database_name: &database_name "withTransaction-tests"
+collection_name: &collection_name "test"
+
+data: []
+
+# These tests use error codes where the TransientTransactionError label will be
+# applied to the error response for commitTransaction. This will cause the
+# entire transaction to be retried instead of commitTransaction.
+#
+# See: https://github.com/mongodb/mongo/blob/r4.1.6/src/mongo/db/handle_request_response.cpp
+tests:
+  -
+    description: transaction is retried commitTransaction TransientTransactionError (NotMaster)
+    failPoint:
+      configureFailPoint: failCommand
+      mode: { times: 1 }
+      data:
+          failCommands: ["commitTransaction"]
+          errorCode: 10107 # NotMaster
+          closeConnection: false
+    operations:
+      -
+        name: withTransaction
+        object: session0
+        callback: &callback
+          operations:
+            -
+              name: insertOne
+              object: collection
+              arguments:
+                session: session0
+                document: { _id: 1 }
+              result:
+                insertedId: 1
+        result:
+          errorCodeName: NotMaster
+          errorLabelsContain: ["TransientTransactionError"]
+          errorLabelsOmit: ["UnknownTransactionCommitResult"]
+    expectations: &expectations
+      -
+        command_started_event:
+          command:
+            insert: *collection_name
+            documents:
+              - { _id: 1 }
+            ordered: true
+            lsid: session0
+            txnNumber: { $numberLong: "1" }
+            startTransaction: true
+            autocommit: false
+            # omitted fields
+            readConcern: ~
+            writeConcern: ~
+          command_name: insert
+          database_name: *database_name
+      -
+        command_started_event:
+          command:
+            commitTransaction: 1
+            lsid: session0
+            txnNumber: { $numberLong: "1" }
+            autocommit: false
+            # omitted fields
+            readConcern: ~
+            startTransaction: ~
+            writeConcern: ~
+          command_name: commitTransaction
+          database_name: admin
+      -
+        command_started_event:
```

Comment:
According to the transaction spec, a NotMaster error is an ordinary retryable write error, and therefore the existing commitTransaction implementation will already retry it without ever having to loop around in withTransaction.  So I'd expect there to be three events here: insert, commitTransaction, and an identical commitTransaction as the last one.

I think most of the tests in this file (and maybe a few others) have this problem.

Reply:
This makes sense. I think the correct fix is to bump the fail point's `mode.times` up to 2 so we fail on both the original attempt and the internal retry attempt. Control would then be returned to `withTransaction()` and we can repeat the entire transaction and succeed with the insert and commitTransaction commands (no retries needed).

Also, since `withTransaction()` ultimately succeeds after the retry, I don't think I should be expecting an error at all. That jives with what you mentioned in https://github.com/mongodb/specifications/pull/435#discussion_r245122170.

While we'll no longer be asserting details about the error, I do want to confirm that we rightfully expect a [TransientTransactionError](https://github.com/mongodb/specifications/blob/master/source/transactions/transactions.rst#transienttransactionerror) label on the `commitTransaction` command result, since it's added by the server (we don't request that the fail point close the connection). [UnknownTransactionCommitResult](https://github.com/mongodb/specifications/blob/master/source/transactions/transactions.rst#unknowntransactioncommitresult) is only ever added by the driver in certain cases (e.g. network, server selection errors), none of which apply here since we do receive a server response.

Reply:
Not working the way you expected.  "UnknownTransactionCommitResult" is indeed being added, and so WithTransaction just retries the commit rather than the whole transaction.  Here's the Java driver's implementation that is adding that label: https://github.com/mongodb/mongo-java-driver/blob/757c3279bb08738e2209d38285b18a9ef3d0ab35/driver-core/src/main/com/mongodb/operation/CommitTransactionOperation.java#L75-L100

See https://github.com/mongodb/specifications/blob/master/source/transactions/transactions.rst#unknowntransactioncommitresult for the transactions spec section that this corresponds to.

Reply:
The tests below that use error codes 79 and 100 should exercise the code path that retries the whole transaction after a server error response to commitTransaction.

Reply:
Should be fixed by https://github.com/mongodb/specifications/pull/435/commits/c09a0c5c81e970b8539a05d1cb3a55a31e1b3be5.